### DISCUSSION TITLE
App Update Helper

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Labs-Windows#128] [Ideas]
I guess I've mis-understood how 'mandatory package updates' worked. I thought it was store enforced, but it's just a flag your app can check for.

You need code to enforce this. See the docs: https://docs.microsoft.com/windows/uwp/packaging/self-install-package-updates#mandatory-package-updates

Maybe it could be helpful to have some sort of helper or custom update dialog (or combination) to help here?

Thoughts?

**Comment by Arlodotexe** [comment]
Seems I've also misunderstood how mandatory package updates worked.

Looks like it's entirely up to the app to handle what to do when a mandatory update exists. The code in the docs is a little messy, but I'm not sure what helpers or controls we could provide without restricting the flexibility of just using the Windows APIs.

Did you have something specific in mind?

**Reply by michael-hawker** [reply]
No not really, I just realized I misunderstood how this worked too, as I thought it was all automatic. Just wanted to call this out here as a potential area for someone to pick up in the future and investigate. Not sure how things may differ between UWP and Windows App SDK. But considering the docs example was not just a line or two of code, it seemed like something that could be useful to have a helper for.

**Reply by Arlodotexe** [reply]
I think there's a lot of room for cleanup and improvement in the docs example, code wise.

I was going to suggest something like a `PackageUpdater` helper, but checking the docs, this already seems pretty easy to do.

From [Package.CheckUpdateAvailabilityAsync](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.package.checkupdateavailabilityasync?view=winrt-22000):

```csharp
private async void CheckForUpdatesButton_Click(object sender, RoutedEventArgs e)
{
    // Get the current app's package for the current user.
    PackageManager pm = new PackageManager();
    Package currentPackage = pm.FindPackageForUser(string.Empty, Package.Current.Id.FullName);

    PackageUpdateAvailabilityResult result = await currentPackage.CheckUpdateAvailabilityAsync();
    switch (result.Availability)
    {
        case PackageUpdateAvailability.Available:
            GoToUpdateAvailableUIView();
            break;
        case PackageUpdateAvailability.Required:
            GoToUpdateRequiredUIView();
            break;
        case PackageUpdateAvailability.NoUpdates:
            // Dismissable ‘Ok’ dialog.
            ShowNoUpdateAvailableDialog(); 
            break;
        case PackageUpdateAvailability.Unknown:
        default:
            // Log and ignore error.
            Logger.Log($"No update information associated with app {Package.Current.DisplayName}");
            // Dismissable ‘Ok’ dialog.
            ShowNoUpdateAvailableDialog();
            break;
    }
}
```

**Reply by michael-hawker** [reply]
I think it was also the code I saw around the `Required` update and having to pull from the store registry, showing the percent, restarting the app, and all that stuff.

Like wondering if we could have a special `ContentDialog` or control helper which encapsulates showing messages to the user and then handling installing the required update if there is one. Then a developer can just add it to their main shell and call it to perform a check either silently or via user interaction...

Main issue would be providing localization for all that sort of text required which isn't something we've done in the past a lot of.

**Reply by Arlodotexe** [reply]
I like the sound of that 🙂

**Reply by mrlacey** [reply]
It would be good to also package the logic so a developer can use their own dialog if they wanted too.
